internal/worker/cache.go

```diff
@@ -0,0 +1,326 @@
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	delete(c.items, key)
+}
+
+type (
+	Bus interface {
+		Contracts(ctx context.Context, opts api.ContractsOpts) ([]api.ContractMetadata, error)
+		GougingParams(ctx context.Context) (api.GougingParams, error)
+		RegisterWebhook(ctx context.Context, wh webhooks.Webhook) error
+	}
+
+	WorkerCache interface {
+		DownloadContracts(ctx context.Context) ([]api.ContractMetadata, error)
+		GougingParams(ctx context.Context) (api.GougingParams, error)
+		HandleEvent(event webhooks.Event) error
+		Initialize(ctx context.Context) error
+	}
+)
+
+type cache struct {
+	b         Bus
+	eventsURL string
+
+	cache  *memoryCache
+	logger *zap.SugaredLogger
+
+	mu    sync.Mutex
+	ready bool
+}
+
+func NewCache(b Bus, workerAddr string, logger *zap.Logger) WorkerCache {
+	return &cache{
+		b:         b,
+		eventsURL: workerAddr,
+
+		cache:  newMemoryCache(),
+		logger: logger.Sugar().Named("workercache"),
+	}
+}
+
+func (c *cache) DownloadContracts(ctx context.Context) (contracts []api.ContractMetadata, err error) {
+	// fetch directly from bus if the cache is not ready
+	if !c.isReady() {
+		c.logger.Warn(errCacheNotReady)
+		contracts, err = c.b.Contracts(ctx, api.ContractsOpts{})
+		return
+	}
+
+	// fetch from bus if it's not cached or expired
+	value, found, expired := c.cache.Get(cacheKeyDownloadContracts)
+	if !found || expired {
+		contracts, err = c.b.Contracts(ctx, api.ContractsOpts{})
+		if err == nil {
+			c.cache.Set(cacheKeyDownloadContracts, contracts)
+		}
+		if expired && !contractsEqual(value.([]api.ContractMetadata), contracts) {
+			c.logger.Warn(fmt.Errorf("%w: key %v", errCacheOutdated, cacheKeyDownloadContracts))
+		}
+		return
+	}
+
+	return value.([]api.ContractMetadata), nil
+}
+
+func (c *cache) GougingParams(ctx context.Context) (gp api.GougingParams, err error) {
+	// fetch directly from bus if the cache is not ready
+	if !c.isReady() {
+		c.logger.Warn(errCacheNotReady)
+		gp, err = c.b.GougingParams(ctx)
+		return
+	}
+
+	// fetch from bus if it's not cached or expired
+	value, found, expired := c.cache.Get(cacheKeyGougingParams)
+	if !found || expired {
+		gp, err = c.b.GougingParams(ctx)
+		if err == nil {
+			c.cache.Set(cacheKeyGougingParams, gp)
+		}
+		if expired && !gougingParamsEqual(value.(api.GougingParams), gp) {
+			c.logger.Warn(fmt.Errorf("%w: key %v", errCacheOutdated, cacheKeyGougingParams))
+		}
+		return
+	}
+
+	return value.(api.GougingParams), nil
+}
+
+func (c *cache) HandleEvent(event webhooks.Event) error {
+	parsed, err := api.ParseEvent(event)
+	if err != nil {
+		return err
+	}
+
+	switch e := parsed.(type) {
+	case api.EventConsensusUpdate:
+		return c.handleConsensusUpdate(e)
+	case api.EventContractArchive:
+		return c.handleContractArchive(e)
+	case api.EventContractRenew:
+		return c.handleContractRenew(e)
+	case api.EventSettingUpdate:
+		return c.handleSettingUpdate(e)
+	case api.EventSettingDelete:
+		return c.handleSettingDelete(e)
```

Comment:
For each of these cases could you add info logging with some context on the event? Just so that if we ever stop getting updates, we know when we received the last one. I think info level is probably ok since it's rare enough.